tests/run/binop_reverse_methods_GH2056.pyx

```diff
@@ -4,29 +4,55 @@ cimport cython
 @cython.cclass
 class Base(object):
     """
-    >>> Base() + 2
+    >>> Base(1) + 2
     'Base.__add__(Base(), 2)'
-    >>> 2 + Base()
+    >>> 2 + Base(1)
     'Base.__radd__(Base(), 2)'
 
-    >>> Base() ** 2
+    >>> Base(0) + 2  #doctest: +ELLIPSIS
+    Traceback (most recent call last):
+    ...
+    TypeError: unsupported operand type...
+    >>> 2 + Base(0)  #doctest: +ELLIPSIS
+    Traceback (most recent call last):
+    ...
+    TypeError: unsupported operand type...
+
+    >>> Base(1) ** 2
     'Base.__pow__(Base(), 2, None)'
-    >>> 2 ** Base()
+    >>> 2 ** Base(1)
     'Base.__rpow__(Base(), 2, None)'
-    >>> pow(Base(), 2, 100)
+    >>> pow(Base(1), 2, 100)
     'Base.__pow__(Base(), 2, 100)'
     """
+    implemented: cython.int
+
+    def __init__(self, implemented):
+        self.implemented = implemented
+
     def __add__(self, other):
-        return "Base.__add__(%s, %s)" % (self, other)
+        if (<Base>self).implemented:
```

Comment:
Isn't the whole point of this feature to make the cast unnecessary?

Reply:
Maybe? https://github.com/cython/cython/issues/2056#issuecomment-656096798

I think `__radd__` could always have the type set correctly. For `__add__` it's more complicated because we're currently going for hybrid behaviour so that old code keeps working without changes.

I think it's probably outside the scope of this PR anyway (which is mainly just aiming to stop valid code from crashing)

Reply:
Yeah, I keep being outsmarted by my past self... Ok, then it doesn't work yet without the cast. Not part of this PR.

Reply:
BTW, when we implement this, I think we should do it for both `__radd__` *and* `__add__` etc. It's ok if older Cython versions don't know about the type of `self`, but new ones should. I don't think this change would hurt existing code.

Reply:
I think the problem for existing code is:

```
cdef class C:
   def __add__(lhs, rhs):
      if isinstance(lhs, C):
         return (<C>lhs).some_attribute
      else:
         return (<C>rhs).some_attribute # but it's not safe to treat lhs as a C
```

it isn't clear how to handle that. Options might be:
* opt-in compiler directive to get "full" proper behavious
* detect `__radd__`, because that makes it clear how `__add__` should behave too.
* detect `isinstance(..., C)` checks inside `__add__` (but that could be unreliable - there's other ways of doing the same thing)

Reply:
Can't we base it on the same (translation time) decisions that allow us to call `__add__` vs. `__radd__` in the first place? If we know that we need our dispatch mechanism, then we know that the first argument is `self`, don't we?

Reply:
I don't think so. The difficulty is when you have a class that only defines `__add__`:

```
class C:
   def __add__(self, other):
      return self
```

In Python `1+C()` is not implemented, so fails.

In Cython it isn't clear if the user wants the old-style version of `C.__add__(1, C())` or the new style version that matches Python and fails.

Obviously there's a lot of cases where it is clear (based on the dispatch mechanism) and we can happily set the types. But the one ambiguous case is probably pretty common.

Reply:
I'm questioning that this is a problem, though. All we'd be doing is set the type of the `self` argument, iff `c_api_binop_methods` is false (the new default). The only problem with this that I can see is if users try to re-assign something to `self` which does not have the right type. To me, that doesn't feel like a common thing to do.

Specifically, the example that you gave [above](https://github.com/cython/cython/pull/4204#discussion_r649352010) is entirely fine, whether `self` is typed or not.

Reply:
I can come up with a contrived example of something that would work in Cython 0.29.x, but break if `self` is automatically typed:

```
class RegularClass:
   def __init__(self):
       self.a = 10

cdef class CdefClass:
    cdef int a
    def __add__(self, other):
       if isinstance(self, CdefClass):
           return "self was CdefClass"
       else:
            return self.a  # should do a Python attribute lookup, but tries to access CdefClass.a directly
```

It's very contrived though - I take the point that it's most likely fine for the common cases (and that there's a simple directive that fixes the code again, and that 3.0 is the time to break existing code if ever, and that we do emit a warning if they only define `__xxx__` and not `__rxxx__` and don't set `c_api_binop_methods`).

So maybe it's worth doing and accepting very occasional problems?